cmd/kube-rbac-proxy/app/options/proxyoptions.go

```diff
@@ -125,13 +138,96 @@ func (o *ProxyOptions) ApplyTo(c *server.KubeRBACProxyInfo, a *serverconfig.Auth
 		}
 	}
 
+	c.UpstreamHeaders = o.UpstreamHeader
 	c.IgnorePaths = o.IgnorePaths
 	c.AllowPaths = o.AllowPaths
 	a.APIAudiences = o.TokenAudiences
 
 	return nil
 }
 
+func validateSecureConnectionConfig(o *ProxyOptions) error {
+	if !identityheaders.HasIdentityHeadersEnabled(o.UpstreamHeader) || !o.UpstreamForceH2C {
+		return nil
+	}
+
+	errLoopback := validateLoopbackAddress(o.Upstream)
+	if errLoopback == nil {
+		return nil
+	}
+	if o.UpstreamForceH2C {
+		return fmt.Errorf("loopback address is required for h2c: %w", errLoopback)
+	}
+
+	errMTLS := validateMTLSConfiguration(o.UpstreamClientCertFile, o.UpstreamClientKeyFile, o.UpstreamCAFile)
+	if errMTLS == nil {
+		return nil
+	}
+
+	return fmt.Errorf(
+		"loopback address or mTLS is required for identity headers: loopback - %w; mTLS - %w",
+		errLoopback,
+		errMTLS,
+	)
+}
+
+func validateMTLSConfiguration(upstreamClientCertPath, upstreamClientKeyPath, upstreamCAPath string) error {
+	// Check if client is configured to provide a certificiate.
+	if len(upstreamClientCertPath) == 0 || len(upstreamClientKeyPath) == 0 || len(upstreamCAPath) == 0 {
+		return fmt.Errorf("krp client cert/key and an upstream CA are required for mTLS")
```

Comment:
Why do you need a special CA for that?

Note that, contrary to the docs, this is not required in that case:
https://github.com/brancz/kube-rbac-proxy/blob/f5d9a22984cf08a7b16b18df8cb3343d9169e7f6/pkg/server/config.go#L84-L97

Reply:
If we have a mTLS setup, the krp client needs to have certs: check.
If we have a mTLS setup, we need a CA from the upstream server we are connecting to.

Reply:
>  If we have a mTLS setup, we need a CA from the upstream server we are connecting to.
unless it's already trusted by the system

Reply:
```mermaidjs
sequenceDiagram
  participant krp
  participant upstream
  
  krp ->> upstream: look, this is my upstreamClientCert and I can encrypt it with upstreamClientKey.
  upstream ->> krp: wow, so cool. Btw, here is my serving certs signed by upstreamCA, you know, I am serving HTTPS.
```